onecodex/models/collection.py

```diff
@@ -389,6 +391,171 @@ def taxonomy(self):
+
+        return self._cached["functional_profiles"]
+
+    def _collate_functional_results(
+        self, annotation, metric, taxa_stratified=True, fill_missing=False, filler=0
+    ):
+        """
+        Return a dataframe of all functional profile data.
+
+        Parameters
+        ----------
+        annotation : {'pathways', 'metacyc', 'eggnog', 'go', 'ko', 'ec', 'pfam', 'reaction'}, optional
+            Annotation data to return
+        taxa_stratified : 'bool', optional
+            Return taxonomically stratified data
+        metric : {'coverage', 'abundance'} for annotation=='pathways', {'rpk', 'cpm'} for other annotations
+            Metric values to return
+        fill_missing : 'bool', optional
+            Fill np.nan values
+        filler : 'float', optional
+            Value with which to fill np.nans
+        """
+        # validate args
+        valid_annotation_values = [
+            "pathways",
+            "metacyc",
+            "eggnog",
+            "go",
+            "ko",
+            "ec",
+            "pfam",
+            "reaction",
+        ]
+        if annotation not in valid_annotation_values:
+            raise ValueError(f"'annotation' must be one of {', '.join(valid_annotation_values)}")
+        if annotation == "pathways":
+            if metric not in ["coverage", "abundance"]:
+                raise ValueError(
+                    "if using annotation='pathways', 'value' must be one of ['coverage', 'abundance']"
+                )
+        elif metric not in ["cpm", "rpk"]:
+            raise ValueError(
+                f"if using annotation={annotation}, 'value' must be one of ['cpm', 'rpk']"
+            )
+
+        data = {}
+        all_features = set()
+        # iterate over functional profiles, subset data, and store in data dict
+        for profile in self._functional_profiles:
+            # get table from mainline
+            table = profile.table(annotation=annotation, taxa_stratified=taxa_stratified)
+            # filter by indicated metric
+            if not taxa_stratified:
+                table = table[table["metric"] == "total_" + metric]
+            else:
+                table = table[table["metric"] == metric]
+            # if taxa stratified, concatenate id and taxon_name
+            if taxa_stratified:
+                table["id"] = table["id"] + "_" + table["taxon_name"]
+            # store tables for later retrieval
+            data[profile.id] = dict(zip(table.id, table.value))
+            all_features.update(set(table["id"]))
+
+        features_to_ix = {}
+        feature_list = []
+        for ix, feature in enumerate(all_features):
+            features_to_ix[feature] = ix
+            feature_list.append(feature)
+
+        # initialize an array and fill it
+        array = np.full(shape=(len(data), len(features_to_ix)), fill_value=np.nan)
+        profile_ids = []
+        for profile_index, profile_id in enumerate(data):
```

Comment:
this is a theoretical concern: you're using the item insert order for `data` dict to determine the order of functional profiles in the result. In Python >=3.5, this will work because `dict` maintains insertion order. But it's usually a good idea to use `OrderedDict` if you are relying on dictionary order. In this case, you could use `self._functional_profiles` list to determine the row order in the dataframe

see https://gandenberger.org/2018/03/10/ordered-dicts-vs-ordereddict/

Reply:
Is your concern specifically that the order of the rows in the returned dataframe may not match the order of the functional profiles in `self._functional_profiles()` (if we're using python <=3.5)? Or is it that there is possibly a data integrity issue? The index of the df is set by `profile_ids`, generated inside the enumeration of `data`, so if I'm reading it correctly, the data table should be intact, though rows could be in a different order than the profiles in `._functional_profiles()`.

Reply:
yeah I was concerned the order of rows in the df *could* differ from the order of `self._functional_profiles`. Not sure if that'll cause any confusion/issues downstream, I just remember that the ordering of classifications matches the df's row order in `_collate_results()` and thought it might be safer to do that here too. Feel free to ignore because it's probably just a theoretical issue or a non-issue